File: go/files.go
package main

import (
	"archive/zip"
	"bufio"
	"encoding/json"
	"encoding/xml"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
)

// Структура для работы с JSON и XML
type Example struct {
	Name  string `json:"name" xml:"name"`
	Value int    `json:"value" xml:"value"`
}

func waitForEnter() {
	fmt.Println("\nНажмите Enter для продолжения...")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
}

func listDisks() {
	fmt.Println("Информация о логических дисках:")
	switch runtime.GOOS {
	case "windows":
		cmd := "wmic logicaldisk get caption,volumename,size,filesystem"
		output, err := exec.Command("cmd", "/C", cmd).Output()
		if err != nil {
			fmt.Println("Ошибка получения информации о дисках:", err)
			return
		}
		fmt.Println(string(output))
	case "darwin":
		cmd := "df -h"
		output, err := exec.Command("sh", "-c", cmd).Output()
		if err != nil {
			fmt.Println("Ошибка получения информации о дисках:", err)
			return
		}
		fmt.Println(string(output))
	default:
		fmt.Println("Функция не поддерживается для этой операционной системы")
	}
}

func createFile(fileName string) {
	fmt.Println("Введите строку для записи в файл:")
	reader := bufio.NewReader(os.Stdin)
	content, _ := reader.ReadString('\n')

	err := ioutil.WriteFile(fileName, []byte(content), 0644)
	if err != nil {
		fmt.Println("Ошибка создания файла:", err)
	} else {
		fmt.Println("Файл успешно создан:", fileName)
	}
}

func readFile(fileName string) {
	data, err := ioutil.ReadFile(fileName)
	if err != nil {
		fmt.Println("Ошибка чтения файла:", err)
		return
	}
	fmt.Println("Содержимое файла:", string(data))
}

func deleteFile(fileName string) {
	err := os.Remove(fileName)
	if err != nil {
		fmt.Println("Ошибка удаления файла:", err)
	} else {
		fmt.Println("Файл успешно удалён:", fileName)
	}
}

func createJSON(fileName string) {
	fmt.Println("Введите имя:")
	var name string
	fmt.Scanln(&name)

	fmt.Println("Введите значение:")
	var value int
	fmt.Scanln(&value)

	exampleJSON := Example{Name: name, Value: value}

	file, err := os.Create(fileName)
	if err != nil {
		fmt.Println("Ошибка создания JSON файла:", err)
		return
	}
	defer file.Close()

	encoder := json.NewEncoder(file)
	err = encoder.Encode(exampleJSON)
	if err != nil {
		fmt.Println("Ошибка записи в JSON файл:", err)
	} else {
		fmt.Println("JSON файл успешно создан:", fileName)
	}
}

func readJSON(fileName string) {
	data, err := ioutil.ReadFile(fileName)
	if err != nil {
		fmt.Println("Ошибка чтения JSON файла:", err)
		return
	}
	fmt.Println("Содержимое JSON файла:", string(data))
}

func createXMLFromFile(fileName string) {
	file, err := os.Create(fileName)
	if err != nil {
		fmt.Println("Ошибка создания XML файла:", err)
		return
	}
	defer file.Close()

	fmt.Println("Введите имя:")
	var name string
	fmt.Scanln(&name)

	fmt.Println("Введите значение:")
	var value int
	fmt.Scanln(&value)

	exampleXML := Example{Name: name, Value: value}

	encoder := xml.NewEncoder(file)
	err = encoder.Encode(exampleXML)
	if err != nil {
		fmt.Println("Ошибка записи в XML файл:", err)
	} else {
		fmt.Println("XML файл успешно создан:", fileName)
	}
}

func readXML(fileName string) {
	data, err := ioutil.ReadFile(fileName)
	if err != nil {
		fmt.Println("Ошибка чтения XML файла:", err)
		return
	}
	fmt.Println("Содержимое XML файла:", string(data))
}

func createZipArchive(zipFileName string, fileName string) {
	zipFile, err := os.Create(zipFileName)
	if err != nil {
		fmt.Println("Ошибка создания ZIP файла:", err)
		return
	}
	defer zipFile.Close()

	zipWriter := zip.NewWriter(zipFile)
	defer zipWriter.Close()

	fileToZip, err := os.Open(fileName)
	if err != nil {
		fmt.Println("Ошибка открытия файла:", err)
		return
	}
	defer fileToZip.Close()

	info, err := fileToZip.Stat()
	if err != nil {
		fmt.Println("Ошибка получения информации о файле:", err)
		return
	}

	header, err := zip.FileInfoHeader(info)
	if err != nil {
		fmt.Println("Ошибка создания заголовка архива:", err)
		return
	}

	header.Name = fileName
	writer, err := zipWriter.CreateHeader(header)
	if err != nil {
		fmt.Println("Ошибка создания записи в архиве:", err)
		return
	}

	_, err = io.Copy(writer, fileToZip)
	if err != nil {
		fmt.Println("Ошибка копирования содержимого файла в архив:", err)
		return
	}

	fmt.Println("ZIP архив успешно создан:", zipFileName)
}

func unzipArchive(zipFileName string, destDir string) {
	zipReader, err := zip.OpenReader(zipFileName)
	if err != nil {
		fmt.Println("Ошибка открытия архива:", err)
		return
	}
	defer zipReader.Close()

	for _, file := range zipReader.File {
		path := filepath.Join(destDir, file.Name)
		if file.FileInfo().IsDir() {
			os.MkdirAll(path, file.Mode())
		} else {
			if err := os.MkdirAll(filepath.Dir(path), file.Mode()); err != nil {
				fmt.Println("Ошибка создания директории для файла:", err)
				return
			}

			outFile, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, file.Mode())
			if err != nil {
				fmt.Println("Ошибка создания файла для извлечения:", err)
				return
			}
			defer outFile.Close()

			rc, err := file.Open()
			if err != nil {
				fmt.Println("Ошибка открытия файла из архива:", err)
				return
			}
			defer rc.Close()

			_, err = io.Copy(outFile, rc)
			if err != nil {
				fmt.Println("Ошибка извлечения файла из архива:", err)
				return
			}

			fmt.Println("Файл успешно извлечён из архива:", path)
		}
	}
}

func main() {
	// Информация о дисках (только для Windows)
	listDisks()
	waitForEnter()

	// Работа с файлами
	fmt.Println("\nРабота с файлами:")
	fileName := "example.txt"
	createFile(fileName)
	waitForEnter()

	readFile(fileName)
	waitForEnter()

	deleteFile(fileName)
	waitForEnter()

	// Работа с JSON
	fmt.Println("\nРабота с JSON:")
	jsonFileName := "example.json"
	createJSON(jsonFileName)
	waitForEnter()

	readJSON(jsonFileName)
	waitForEnter()

	deleteFile(jsonFileName)
	waitForEnter()

	// Работа с XML
	fmt.Println("\nРабота с XML:")
	xmlFileName := "example.xml"
	createXMLFromFile(xmlFileName)
	waitForEnter()

	readXML(xmlFileName)
	waitForEnter()

	deleteFile(xmlFileName)
	waitForEnter()

	// Работа с ZIP архивом
	fmt.Println("\nРабота с ZIP архивом:")
	fileToArchive := "file_for_zip.txt"
	createFile(fileToArchive)
	waitForEnter()

	zipFileName := "archive.zip"
	createZipArchive(zipFileName, fileToArchive)
	waitForEnter()

	unzipArchive(zipFileName, "./unzipped")
	waitForEnter()

	deleteFile(fileToArchive)
	deleteFile(zipFileName)
	waitForEnter()
}

File: python/files.py
import os
import json
import xml.etree.ElementTree as ET
import zipfile
import platform

class Example:
    def __init__(self, name, value):
        self.name = name
        self.value = value

def wait_for_enter():
    input("\nPress Enter to continue...")

def list_disks():
    print("Information about logical disks:")
    if platform.system() == "Windows":
        os.system("wmic logicaldisk get caption,volumename,size,filesystem")
    elif platform.system() == "Darwin":  # macOS
        os.system("df -h")
    else:
        print("Function not supported for this operating system")

def create_file(file_name):
    content = input("Enter a string to write to the file: ")
    with open(file_name, 'w') as file:
        file.write(content)
    print(f"File successfully created: {file_name}")

def read_file(file_name):
    with open(file_name, 'r') as file:
        data = file.read()
    print(f"Contents of the file:\n{data}")

def delete_file(file_name):
    try:
        os.remove(file_name)
        print(f"File successfully deleted: {file_name}")
    except Exception as e:
        print(f"Error deleting file: {e}")

def create_json(file_name):
    name = input("Enter name: ")

    while True:
        try:
            value = int(input("Enter value (integer only): "))
            break  # Break the loop if the conversion is successful
        except ValueError:
            print("Invalid input. Please enter an integer.")

    example_json = Example(name, value).__dict__
    
    with open(file_name, 'w') as file:
        json.dump(example_json, file, indent=4)
    print(f"JSON file successfully created: {file_name}")

def read_json(file_name):
    with open(file_name, 'r') as file:
        data = file.read()
    print(f"Contents of JSON file:\n{data}")

def create_xml(file_name):
    name = input("Enter name: ")
    
    while True:
        try:
            value = int(input("Enter value (integer only): "))
            break  # Break the loop if the conversion is successful
        except ValueError:
            print("Invalid input. Please enter an integer.")

    example_xml = Example(name, value)

    root = ET.Element("Example")
    ET.SubElement(root, "name").text = example_xml.name
    ET.SubElement(root, "value").text = str(example_xml.value)

    tree = ET.ElementTree(root)
    tree.write(file_name, xml_declaration=True, encoding='utf-8')
    print(f"XML file successfully created: {file_name}")

def read_xml(file_name):
    tree = ET.parse(file_name)
    root = tree.getroot()
    print(f"Contents of XML file:\n{ET.tostring(root, encoding='utf-8').decode('utf-8')}")

def create_zip_archive(zip_file_name, file_name):
    with zipfile.ZipFile(zip_file_name, 'w') as zipf:
        zipf.write(file_name, os.path.basename(file_name))
    print(f"ZIP archive successfully created: {zip_file_name}")

def unzip_archive(zip_file_name, dest_dir):
    with zipfile.ZipFile(zip_file_name, 'r') as zipf:
        zipf.extractall(dest_dir)
    print(f"Files successfully extracted to: {dest_dir}")

def main():
    # List disks (only for Windows and macOS)
    list_disks()
    wait_for_enter()

    # File operations
    print("\nFile operations:")
    file_name = "example.txt"
    create_file(file_name)
    wait_for_enter()

    read_file(file_name)
    wait_for_enter()

    delete_file(file_name)
    wait_for_enter()

    # JSON operations
    print("\nJSON operations:")
    json_file_name = "example.json"
    create_json(json_file_name)
    wait_for_enter()

    read_json(json_file_name)
    wait_for_enter()

    delete_file(json_file_name)
    wait_for_enter()

    # XML operations
    print("\nXML operations:")
    xml_file_name = "example.xml"
    create_xml(xml_file_name)
    wait_for_enter()

    read_xml(xml_file_name)
    wait_for_enter()

    delete_file(xml_file_name)
    wait_for_enter()

    # ZIP archive operations
    print("\nZIP archive operations:")
    file_to_archive = "file_for_zip.txt"
    create_file(file_to_archive)
    wait_for_enter()

    zip_file_name = "archive.zip"
    create_zip_archive(zip_file_name, file_to_archive)
    wait_for_enter()

    unzip_archive(zip_file_name, "./unzipped")
    wait_for_enter()

    delete_file(file_to_archive)
    delete_file(zip_file_name)
    wait_for_enter()

if __name__ == "__main__":
    main()